enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,168 @@
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an unlimited history of remediations.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was succesfully fenced (deleted / powered off)
+- when the node is healthy again.
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+Example MHC:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineHealthCheck
+metadata:
+  annotations:
+    machine.openshift.io/remediation-strategy: external-baremetal
+  creationTimestamp: "2020-11-24T16:31:22Z"
+spec:
+  maxUnhealthy: 100%
+  nodeStartupTimeout: 60m
+  selector:
+    matchLabels:
+      machine.openshift.io/cluster-api-machine-role: worker
+  unhealthyConditions:
+  - status: "False"
+    timeout: 20s
+    type: Ready
+  - status: Unknown
+    timeout: 20s
+    type: Ready
+status:
+  conditions:
+  - lastTransitionTime: "2020-11-24T17:17:32Z"
+    status: "True"
+    type: RemediationAllowed
+  currentHealthy: 2
+  expectedMachines: 2
+  remediationHistory:
+  - conditionStatus: Unknown
+    conditionType: Ready
+    detected: "2020-11-24T17:33:43Z"
+    fenced: "2020-11-24T17:34:35Z"
+    finished: "2020-11-24T17:35:09Z"
+    remediationType: external
+    started: "2020-11-24T17:34:04Z"
+    targetKind: Node
+    targetName: worker-1
+  remediationsAllowed: 2
+```
+
+Example UI: TBD (copy from "KNI Node & Host Management Designs" doc?)
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator I want to have an overview over when and why nodes failed and were remediated, in order
+to detect patterns, and to be able to troubleshoot possible root causes.
```

Comment:
The why often has to be collected elsewhere.  MHC for example simply knows a machine was in status Unknown for many types of scenarios.  Correlating to resource consumption may help but not always.  We have a specific example in the OSD support queue right now that we're hitting a dead end tracking a root cause on because the machine is gone, we can't see what actually was happening.